rasa/core/training/story_conflict.py

```diff
@@ -0,0 +1,242 @@
+from collections import defaultdict, namedtuple
+from typing import List, Optional, Dict, Text, Tuple, Generator, NamedTuple
+
+from rasa.core.actions.action import ACTION_LISTEN_NAME
+from rasa.core.domain import PREV_PREFIX, Domain
+from rasa.core.events import ActionExecuted, Event
+from rasa.core.featurizers import MaxHistoryTrackerFeaturizer
+from rasa.nlu.constants import INTENT_ATTRIBUTE
+from rasa.core.training.generator import TrackerWithCachedStates
+
+
+class StoryConflict:
+    def __init__(self, sliced_states: List[Optional[Dict[Text, float]]],) -> None:
+        self._sliced_states = sliced_states
+        self._conflicting_actions = defaultdict(
+            list
+        )  # {"action": ["story_1", ...], ...}
+
+    def __hash__(self) -> int:
+        return hash(str(list(self._sliced_states)))
+
+    def add_conflicting_action(self, action: Text, story_name: Text) -> None:
+        """Adds another action that follows from the same state.
+
+        Args:
+            action: Name of the action.
+            story_name: Name of the story where this action is chosen.
+        """
+        self._conflicting_actions[action] += [story_name]
+
+    @property
+    def conflicting_actions(self) -> List[Text]:
+        """List of conflicting actions.
+
+        Returns:
+            List of conflicting actions.
+
+        """
+        return list(self._conflicting_actions.keys())
+
+    @property
+    def conflict_has_prior_events(self) -> bool:
+        """Checks if prior events exist.
+
+        Returns:
+            `True` if anything has happened before this conflict, otherwise `False`.
+        """
+        return _get_previous_event(self._sliced_states[-1])[0] is not None
+
+    def __str__(self) -> Text:
+        # Describe where the conflict occurs in the stories
+        last_event_type, last_event_name = _get_previous_event(self._sliced_states[-1])
+        if last_event_type:
+            conflict_message = (
+                f"CONFLICT after {last_event_type} '{last_event_name}':\n"
+            )
+        else:
+            conflict_message = f"CONFLICT at the beginning of stories:\n"
+
+        # List which stories are in conflict with one another
+        for action, stories in self._conflicting_actions.items():
+            conflict_message += "  " + self._summarize_conflict(action, stories)
+
+        return conflict_message
+
+    @staticmethod
+    def _summarize_conflict(action: Text, stories: List[Text]) -> Text:
+        if len(stories) > 3:
+            # Four or more stories are present
+            conflict_description = (
+                f"'{stories[0]}', '{stories[1]}', and {len(stories) - 2} other trackers"
+            )
+        elif len(stories) == 3:
```

Comment:
What do you mean here? This would give a different output. E.g. when only two stories are in the list, then there should be no comma before the `and`.

Reply:
> E.g. when only two stories are in the list, then there should be no comma before the and.

But apart from the `,` ? You also fix the fix case where there is only story.

But okay for me, not a crucial change!